networkx/classes/graph.py

```diff
@@ -365,6 +365,7 @@ def __init__(self, incoming_graph_data=None, **attr):
         self.graph = self.graph_attr_dict_factory()  # dictionary for graph attributes
         self._node = self.node_dict_factory()  # empty node attribute dict
         self._adj = self.adjlist_outer_dict_factory()  # empty adjacency dict
+        self.__networkx_cache__ = {}
```

Comment:
Should we add e.g. `cache_dict_factory` class attribute and use that here just like how all the other dict attributes use factories?

Reply:
My immediate reaction is: No, we should not have a `cache_dict_factory` -- unless we believe that someone would want to replace that dict with a user-grown cache object. I guess we can always imagine that. OK... so why is my reaction like that?

The dict_factories have been a pain to maintain and haven't resulted in the backend kind of behavior we envisioned could occur. Indeed, one could argue that with `_dispatchable` we can now get rid of the dict_factories in the classes. They do serve different purposes, so maybe that argument wouldn't have a clear outcome. But much of the justification for adding them is becoming moot.

I see the cache dict isn easy/quick/flexible data structure, not a fancy data structure. But maybe I'm wrong. Do people see wanting a logging feature on the cache? Or something similar? If so, would the attribute be created more than once? That is, would we need a factory? One reason for the dict_factories is that each node and edge added creates a new dict -- and that is spread out in the code base. Changing the object from a dict ot a special object would need to be done carefully with exhaustive searches because there are many ways to create new dicts and many places in nx where we e.g. create edges.  I foresee us `clear`ing the cache dict in many places, but not creating a new one. There is only a cache for the Graph, not for each edge of the graph.

But I might be missing something. Ideas welcome!